test/tests/functional/pbs_job_array.py

```diff
@@ -744,3 +782,237 @@ def test_recover_big_array_job(self):
+
+        self.create_max_run_subjobs_hook(3, "queuejob", "h1")
+        j1 = Job(attrs={ATTR_J: '1-20'})
+        jid1 = self.server.submit(j1)
+        self.server.log_match("max_run_subjobs set to 3")
+        self.server.expect(JOB, {ATTR_state: 'B'}, id=jid1)
+        self.server.expect(JOB, {'job_state=R': 3}, extend='t')
+
+        # Submit a normal job and see if queuejob hook cannot set the
+        # attribute.
+        with self.assertRaises(PbsSubmitError) as e:
+            self.server.submit(Job())
+        self.assertIn("Not an array job", e.exception.msg[0])
+
+    @skipOnCpuSet
+    def test_max_run_subjobs_modifyjob_hook(self):
+        """
+        Submit array job with large max_run_subjobs limit see if modifyjob
+        modifies it.
+        """
+        a = {'resources_available.ncpus': 20}
+        self.server.manager(MGR_CMD_SET, NODE, a, self.mom.shortname)
+
+        self.create_max_run_subjobs_hook(3, "modifyjob", "h1")
+        self.server.manager(MGR_CMD_SET, SERVER, {'scheduling': 'False'})
+        j = Job(attrs={ATTR_J: '1-50'})
+        jid = self.server.submit(j)
+        self.server.expect(JOB, {ATTR_state: 'Q'}, id=jid)
+        self.server.alterjob(jid, {ATTR_W: 'max_run_subjobs=20'})
+        self.server.log_match("max_run_subjobs set to 3")
+        self.server.manager(MGR_CMD_SET, SERVER, {'scheduling': 'True'})
+        self.server.expect(JOB, {'job_state=R': 3}, extend='t')
+
+        # Modify a normal job and see if queuejob hook cannot set the
+        # attribute.
+        self.server.manager(MGR_CMD_SET, SERVER, {'scheduling': 'False'})
+        nj = self.server.submit(Job())
+        with self.assertRaises(PbsAlterError) as e:
+            self.server.alterjob(nj, {ATTR_W: 'max_run_subjobs=20'})
+        self.assertIn("Not an array job", e.exception.msg[0])
+
+    @skipOnCpuSet
+    def test_max_run_subjobs_preemption(self):
+        """
+        Submit array job with max_run_subjobs limit and see if such a job
+        hits the limit, no preemption is attempted.
+        """
+        a = {'queue_type': 'execution',
+             'started': 'True',
+             'enabled': 'True',
+             'Priority': 200}
+        self.server.manager(MGR_CMD_CREATE, QUEUE, a, "wq2")
+
+        a = {'resources_available.ncpus': 8}
+        self.server.manager(MGR_CMD_SET, NODE, a, self.mom.shortname)
+
+        a = {'Resource_List.select': 'ncpus=2'}
+        j = Job(attrs=a)
+        jid = self.server.submit(j)
+        self.server.expect(JOB, {ATTR_state: 'R'}, id=jid)
+
+        a = {ATTR_J: '1-20%3', 'Resource_List.select': 'ncpus=2',
+             ATTR_q: 'wq2'}
+        j_arr = Job(attrs=a)
+        jid_arr = self.server.submit(j_arr)
+        self.server.expect(JOB, {ATTR_state: 'B'}, id=jid_arr)
+        self.server.expect(JOB, {'job_state=R': 4}, extend='t')
+        self.server.expect(JOB, {ATTR_state: 'R'}, id=jid)
+
+    @skipOnCpuSet
+    def test_max_run_subjobs_qrun(self):
```

Comment:
this test currently checks the admin's ability to qrun
can this also test what happens if the user tries to qrun

Reply:
I think qrun can only be executed by a manager or operator privilege user.

Reply:
oh, you're right